server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -5597,6 +5616,79 @@ public HypervisorType getHypervisorTypeOfUserVM(long vmId) {
         return userVm.getHypervisorType();
     }
 
+    protected String finalizeUserData(String userData, Long userDataId, VirtualMachineTemplate template) {
+        if (StringUtils.isEmpty(userData) && userDataId == null && (template == null || template.getUserDataId() == null)) {
+            return null;
+        }
+        if (template != null && template.getUserDataId() != null) {
+            switch (template.getUserDataOverridePolicy()) {
+                case denyoverride:
+                    if (StringUtils.isNotEmpty(userData) || userDataId != null) {
+                        String msg = String.format("UserData input is not allowed here since template %s is configured to deny any userdata", template.getName());
+                        throw new CloudRuntimeException(msg);
+                    }
+                case allowoverride:
+                    if (userDataId != null) {
+                        if (userData != null) {
+                            s_logger.info("Both userdata and userdata ID are provided, precedence goes to userdata ID");
```

Comment:
would it be better to throw an exception if both userdata and userdataId are provided ?